botocore/compress.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import gzip
+import io
+import logging
+
+from botocore.utils import determine_content_length
+
+logger = logging.getLogger(__name__)
+
+
+def compress_request(config, request_dict, operation_model):
```

Comment:
A bit late in the game to bring this up, but an idea after seeing all the examples in the tests where this doesn't compress: Would it be more clear to call this `maybe_compress_request`? It's a naming pattern I've first encountered in [Smithy repos](https://github.com/smithy-lang) and found very useful for understanding that a function only conditionally does what its name suggest.

Reply:
Yep, I think that's reasonable. I'll throw `attempt_request_compression` as an alternative but I think they both work well.

Reply:
I think I actually called this `maybe_compress_request` in an earlier iteration so I'd vote to change it back to that.